docs/guide/tutorials/task-guide.md

```diff
@@ -0,0 +1,524 @@
+:::
+
+## Task Inputs and Outputs
+In many workflow systems, individual "Tasks" are only allowed to report a minimal set of information about their state (for example, "Finished", "Failed" or "Sucessful").  In Prefect, we encourage Tasks to actually exchange richer information, including "arbitrary" data objects.  
+
+However, there _can_ be restrictions on what tasks can receive as inputs and return as outputs.  In particular, Prefect's most popular executor is the `DaskExecutor`, which allows users to execute tasks on a cluster; because each machine is running a different Python process, Dask must convert Python objects to bytecode that can be shared between different processes. Consequently, the data that is passed around the Prefect platform must be compatible with this serialization protocol.  Specifically, tasks must operate on objects that are serializable by the [`cloudpickle` library](https://github.com/cloudpipe/cloudpickle).
+
+**What can cloudpickle serialize?**
+- most data structures, including lists, sets, and dictionaries
+- functions (including lambda functions)
+- most classes, as long as their attributes are serializable
+
+**What can’t cloudpickle serialize?**
+- generators
+- file objects
+- thread locks
+- deeply recursive classes
+
+**How can I determine if something is serializable?**
+
+Simply import `cloudpickle` and attempt to make the round trip:
+```python
+import cloudpickle
+
+cloudpickle.dumps(None)
+# b'\x80\x04\x95\x02\x00\x00\x00\x00\x00\x00\x00N.'
+
+obj = cloudpickle.loads(cloudpickle.dumps(None))
+assert obj is None
+
+custom_generator = (x for x in range(5))
+cloudpickle.dumps(custom_generator)
+# TypeError: can't pickle generator objects
+```
+
+This subtle and technical constraint actually informs a lot of design decisions in the Task library ([as we will see shortly](#best-practices-in-writing-tasks)). Moreover, users will undoubtedly run into this situation on occasion, and it’s important to recognize the types of errors that arise when this constraint is violated. Also, as you write your own flows, keep this in mind when you design your Tasks and how they communicate information.
+
+## Adding Tasks to Flows
+
+Now that you've written your Tasks, it's time to add them to a Flow.  Sticking with our extra simple `number_task` created above, let's add this task to our Flow using Prefect's Imperative API:
+```python
+from prefect import Flow
+
+f = Flow("example")
+f.add_task(number_task)
+
+print(f.tasks) # {<Task: number_task>}
+```
+
+So far, so good - our Flow now consists of a single task. How might we add a single task to a Flow using the Functional API?  In this instance, we have to perform some _action_ on the Task to "register" it with the Flow.  In general, Tasks will be auto-added to a Flow in the functional API if and only if:
+- the task is _called_ within a Flow context
+- the task is called as a dependency of another task
```

Comment:
Should be specified that one OR the other can happen

Reply:
Yea, good call.  I definitely want to be as clear as possible.